File: ports/go/diff.go
package hqtui

import (
	"strconv"
	"strings"
)

// Turns two framebuffers into the smallest practical stream of escape
// sequences. The encoder keeps a model of the terminal's current pen so no
// redundant SGR is emitted.

// Rewriting up to this many unchanged cells is cheaper than the escape sequence
// needed to jump over them, so neighbouring dirty spans get merged.
const gapMerge = 5

type terminalState struct {
	x, y int
	// known is false after anything that makes the cursor position unknowable.
	known bool
	fg    Color
	bg    Color
	attrs Attrs
}

func freshState() terminalState {
	return terminalState{x: -1, y: -1, known: false, fg: DefaultColor, bg: DefaultColor}
}

type EncodeResult struct {
	// Output is the bytes to hand to stdout.
	Output       string
	ChangedCells int
	// DirtyRows is the number of rows that contained at least one change.
	DirtyRows int
}

type EncoderOptions struct {
	Colors ColorDepth
	// Monochrome drains all color, keeping attributes.
	Monochrome bool
}

type Encoder struct {
	state      terminalState
	parts      strings.Builder
	Colors     ColorDepth
	Monochrome bool
}

func NewEncoder(o EncoderOptions) *Encoder {
	return &Encoder{state: freshState(), Colors: o.Colors, Monochrome: o.Monochrome}
}

// InvalidateState forgets what we believe about the terminal; the next write
// re-states everything.
func (e *Encoder) InvalidateState() {
	e.state = freshState()
	e.parts.WriteString("\x1b[0m")
}

func (e *Encoder) fgSeq(c Color) string {
	if e.Colors == ColorNone {
		return ""
	}
	if c.IsDefault() {
		return AnsiFgDefault
	}
	col := c
	if e.Monochrome {
		col = c.Grayscale()
	}
	switch e.Colors {
	case ColorTrueColor:
		return FgTrue(col.R(), col.G(), col.B())
	case ColorAnsi256:
		return Fg256(col.To256())
	}
	return Fg16(col.To16())
}

func (e *Encoder) bgSeq(c Color) string {
	if e.Colors == ColorNone {
		return ""
	}
	if c.IsDefault() {
		return AnsiBgDefault
	}
	col := c
	if e.Monochrome {
		col = c.Grayscale()
	}
	switch e.Colors {
	case ColorTrueColor:
		return BgTrue(col.R(), col.G(), col.B())
	case ColorAnsi256:
		return Bg256(col.To256())
	}
	return Bg16(col.To16())
}

func (e *Encoder) applyStyle(fg, bg Color, attrs Attrs) {
	s := &e.state
	if s.fg == fg && s.bg == bg && s.attrs == attrs {
		return
	}

	// Attributes can only be added cheaply; removing one means a full reset.
	if removed := s.attrs &^ attrs; removed != 0 {
		e.parts.WriteString("\x1b[0m")
		s.attrs = AttrNone
		s.fg = DefaultColor
		s.bg = DefaultColor
	}

	if added := attrs &^ s.attrs; added != 0 {
		codes := make([]int, 0, 7)
		for _, pair := range []struct {
			bit  Attrs
			code int
		}{
			{AttrBold, 1}, {AttrDim, 2}, {AttrItalic, 3}, {AttrUnderline, 4},
			{AttrBlink, 5}, {AttrReverse, 7}, {AttrStrike, 9},
		} {
			if added&pair.bit != 0 {
				codes = append(codes, pair.code)
			}
		}
		if len(codes) > 0 {
			e.parts.WriteString("\x1b[")
			for i, code := range codes {
				if i > 0 {
					e.parts.WriteByte(';')
				}
				e.parts.WriteString(strconv.Itoa(code))
			}
			e.parts.WriteByte('m')
		}
		s.attrs = attrs
	}

	if s.fg != fg {
		e.parts.WriteString(e.fgSeq(fg))
		s.fg = fg
	}
	if s.bg != bg {
		e.parts.WriteString(e.bgSeq(bg))
		s.bg = bg
	}
}

func (e *Encoder) moveCursor(x, y int) {
	s := &e.state
	if s.known && s.y == y {
		switch {
		case s.x == x:
			return
		case x > s.x && x-s.x <= 3:
			// Short hop: cheaper than a full CUP, and never repaints cells.
			e.parts.WriteString(MoveRight(x - s.x))
		case x == 0:
			e.parts.WriteByte('\r')
		default:
			e.parts.WriteString(MoveToColumn(x))
		}
	} else {
		e.parts.WriteString(MoveTo(x, y))
	}
	s.x, s.y, s.known = x, y, true
}

// Encode writes the difference between prev and next. Pass full to repaint
// every cell (first frame, resize, or after a redraw request).
func (e *Encoder) Encode(prev, next *FrameBuffer, full bool) EncodeResult {
	e.parts.Reset()
	changed, dirtyRows := 0, 0

	w, h := next.Width, next.Height
	sameSize := prev.Width == w && prev.Height == h
	repaint := full || !sameSize
	if repaint {
		e.InvalidateState()
	}

	// differs reports whether a cell actually changed. When the sizes disagree
	// the previous frame cannot be indexed with this frame's stride, so every
	// cell counts as different.
	differs := func(i int) bool {
		if !sameSize {
			return true
		}
		return next.Chars[i] != prev.Chars[i] || next.Fg[i] != prev.Fg[i] ||
			next.Bg[i] != prev.Bg[i] || next.Attrs[i] != prev.Attrs[i]
	}
	// dirtyAt reports whether a cell needs emitting, which a full repaint forces
	// even for cells that are unchanged. The two are deliberately separate:
	// ChangedCells reports real churn, not repaint volume.
	dirtyAt := func(i int) bool { return repaint || differs(i) }

	for y := 0; y < h; y++ {
		rowStart := y * w
		x := 0
		rowDirty := false

		for x < w {
			if !dirtyAt(rowStart + x) {
				x++
				continue
			}

			// Walk left onto the lead cell if we landed on a wide char's tail.
			start := x
			for start > 0 && next.Chars[rowStart+start] == Continuation {
				start--
			}

			// Extend the run while cells are dirty, tolerating short gaps.
			end, clean, probe := start, 0, start
			for probe < w {
				if dirtyAt(rowStart + probe) {
					end, clean = probe, 0
				} else {
					clean++
					if clean > gapMerge {
						break
					}
				}
				probe++
			}

			e.moveCursor(start, y)
			for cx := start; cx <= end; cx++ {
				j := rowStart + cx
				value := next.Chars[j]
				if value == Continuation {
					continue // emitted with its lead cell
				}
				e.applyStyle(next.Fg[j], next.Bg[j], next.Attrs[j])
				if value == 0 {
					e.parts.WriteByte(' ')
				} else {
					e.parts.WriteString(CellText(value))
				}
				e.state.x += max(1, CellWidth(value))
				if differs(j) {
					changed++
				}
			}
			// Writing the final column may have triggered autowrap; stop
			// trusting x.
			if e.state.x >= w {
				e.state.known = false
			}
			rowDirty = true
			x = end + 1
		}
		if rowDirty {
			dirtyRows++
		}
	}

	return EncodeResult{Output: e.parts.String(), ChangedCells: changed, DirtyRows: dirtyRows}
}

// EncodeFull is a one-shot encode of a whole buffer, e.g. for a screenshot.
func EncodeFull(buffer *FrameBuffer, o EncoderOptions) string {
	empty := NewFrameBuffer(buffer.Width, buffer.Height)
	return NewEncoder(o).Encode(empty, buffer, true).Output
}

File: ports/python/hqtui/diff.py
"""Turns two framebuffers into the smallest practical stream of escape sequences.

The encoder keeps a model of the terminal's current pen so no redundant SGR is
emitted.
"""

from __future__ import annotations

from dataclasses import dataclass

from . import ansi
from .buffer import Attrs, FrameBuffer
from .capabilities import ColorDepth
from .color import Color, DEFAULT_COLOR
from .unicode import CONTINUATION, cell_text, cell_width

__all__ = ["EncodeResult", "Encoder", "encode_full"]

_GAP_MERGE = 5
"""Rewriting up to this many unchanged cells is cheaper than the escape sequence
needed to jump over them, so neighbouring dirty spans get merged."""

_ATTR_CODES = (
    (Attrs.BOLD, 1),
    (Attrs.DIM, 2),
    (Attrs.ITALIC, 3),
    (Attrs.UNDERLINE, 4),
    (Attrs.BLINK, 5),
    (Attrs.REVERSE, 7),
    (Attrs.STRIKE, 9),
)


@dataclass(frozen=True, slots=True)
class EncodeResult:
    output: str
    """The bytes to hand to stdout."""
    changed_cells: int
    dirty_rows: int
    """Rows that contained at least one change."""


class Encoder:
    __slots__ = ("colors", "monochrome", "_x", "_y", "_known", "_fg", "_bg", "_attrs", "_parts")

    def __init__(
        self, colors: ColorDepth = ColorDepth.TRUECOLOR, monochrome: bool = False
    ) -> None:
        self.colors = colors
        self.monochrome = monochrome
        self._parts: list[str] = []
        self._reset_state()

    def _reset_state(self) -> None:
        self._x = -1
        self._y = -1
        # False after anything that makes the cursor position unknowable.
        self._known = False
        self._fg = DEFAULT_COLOR
        self._bg = DEFAULT_COLOR
        self._attrs = 0

    def invalidate_state(self) -> None:
        """Forget what we believe about the terminal; the next write re-states
        everything."""
        self._reset_state()
        self._parts.append("\x1b[0m")

    def _fg_seq(self, c: Color) -> str:
        if self.colors == ColorDepth.NONE:
            return ""
        if c == DEFAULT_COLOR:
            return ansi.FG_DEFAULT
        col = c.grayscale() if self.monochrome else c
        if self.colors == ColorDepth.TRUECOLOR:
            return ansi.fg_true(col.r, col.g, col.b)
        if self.colors == ColorDepth.ANSI256:
            return ansi.fg_256(col.to_256())
        return ansi.fg_16(col.to_16())

    def _bg_seq(self, c: Color) -> str:
        if self.colors == ColorDepth.NONE:
            return ""
        if c == DEFAULT_COLOR:
            return ansi.BG_DEFAULT
        col = c.grayscale() if self.monochrome else c
        if self.colors == ColorDepth.TRUECOLOR:
            return ansi.bg_true(col.r, col.g, col.b)
        if self.colors == ColorDepth.ANSI256:
            return ansi.bg_256(col.to_256())
        return ansi.bg_16(col.to_16())

    def _apply_style(self, fg: Color, bg: Color, attrs: int) -> None:
        if self._fg == fg and self._bg == bg and self._attrs == attrs:
            return

        # Attributes can only be added cheaply; removing one means a full reset.
        if self._attrs & ~attrs:
            self._parts.append("\x1b[0m")
            self._attrs = 0
            self._fg = DEFAULT_COLOR
            self._bg = DEFAULT_COLOR

        added = attrs & ~self._attrs
        if added:
            codes = [code for bit, code in _ATTR_CODES if added & bit]
            if codes:
                self._parts.append("\x1b[" + ";".join(str(c) for c in codes) + "m")
            self._attrs = attrs

        if self._fg != fg:
            self._parts.append(self._fg_seq(fg))
            self._fg = fg
        if self._bg != bg:
            self._parts.append(self._bg_seq(bg))
            self._bg = bg

    def _move_cursor(self, x: int, y: int) -> None:
        if self._known and self._y == y:
            if self._x == x:
                return
            if x > self._x and x - self._x <= 3:
                # Short hop: cheaper than a full CUP, and never repaints cells.
                self._parts.append(ansi.move_right(x - self._x))
            elif x == 0:
                self._parts.append("\r")
            else:
                self._parts.append(ansi.move_to_column(x))
        else:
            self._parts.append(ansi.move_to(x, y))
        self._x, self._y, self._known = x, y, True

    def encode(self, prev: FrameBuffer, next_: FrameBuffer, full: bool = False) -> EncodeResult:
        """Encode the difference between ``prev`` and ``next_``.

        Pass ``full`` to repaint every cell (first frame, resize, or after a
        redraw request).
        """
        self._parts = []
        changed = 0
        dirty_rows = 0

        w, h = next_.width, next_.height
        same_size = prev.width == w and prev.height == h
        repaint = full or not same_size
        if repaint:
            self.invalidate_state()

        nc, nf, nb, na = next_.chars, next_.fg, next_.bg, next_.attrs
        pc, pf, pb, pa = prev.chars, prev.fg, prev.bg, prev.attrs

        def differs(i: int) -> bool:
            """Whether this cell actually changed. When the sizes disagree the
            previous frame cannot be indexed with this frame's stride, so every
            cell counts as different."""
            if not same_size:
                return True
            return nc[i] != pc[i] or nf[i] != pf[i] or nb[i] != pb[i] or na[i] != pa[i]

        def dirty_at(i: int) -> bool:
            """Whether this cell needs emitting, which a full repaint forces even
            for unchanged cells. The two are deliberately separate:
            ``changed_cells`` reports real churn, not repaint volume."""
            return repaint or differs(i)

        for y in range(h):
            row_start = y * w
            x = 0
            row_dirty = False

            while x < w:
                if not dirty_at(row_start + x):
                    x += 1
                    continue

                # Walk left onto the lead cell if we landed on a wide char's tail.
                start = x
                while start > 0 and nc[row_start + start] == CONTINUATION:
                    start -= 1

                # Extend the run while cells are dirty, tolerating short gaps.
                end, clean, probe = start, 0, start
                while probe < w:
                    if dirty_at(row_start + probe):
                        end, clean = probe, 0
                    else:
                        clean += 1
                        if clean > _GAP_MERGE:
                            break
                    probe += 1

                self._move_cursor(start, y)
                for cx in range(start, end + 1):
                    j = row_start + cx
                    value = nc[j]
                    if value == CONTINUATION:
                        continue  # emitted with its lead cell
                    self._apply_style(Color(nf[j]), Color(nb[j]), na[j])
                    self._parts.append(" " if value == 0 else cell_text(value))
                    self._x += max(1, cell_width(value))
                    if differs(j):
                        changed += 1
                # Writing the final column may have triggered autowrap; stop
                # trusting x.
                if self._x >= w:
                    self._known = False
                row_dirty = True
                x = end + 1

            if row_dirty:
                dirty_rows += 1

        return EncodeResult("".join(self._parts), changed, dirty_rows)


def encode_full(
    buffer: FrameBuffer,
    colors: ColorDepth = ColorDepth.TRUECOLOR,
    monochrome: bool = False,
) -> str:
    """One-shot encode of a whole buffer, e.g. for a screenshot."""
    empty = FrameBuffer(buffer.width, buffer.height)
    return Encoder(colors, monochrome).encode(empty, buffer, True).output
